flashinfer/comm/trtllm_mnnvl_ar.py

```diff
@@ -0,0 +1,159 @@
+            "buffer_flags_mnnvl",
+            "nranks",
+            "rank",
+            "wait_for_results",
+            "launch_with_pdl",
+        ],
+    )
+    def trtllm_mnnvl_all_reduce(
+        inp: torch.Tensor,
+        out: torch.Tensor,
+        multicast_buffer_ptr: int,  # Pointer address as integer
+        buffer_ptrs_dev: int,  # Pointer address as integer
+        buffer_mnnvl: torch.Tensor,
+        buffer_flags_mnnvl: torch.Tensor,
+        nranks: int,
+        rank: int,
+        wait_for_results: bool,
+        launch_with_pdl: bool,
+    ) -> None:
+        module.trtllm_mnnvl_all_reduce(
+            inp,
+            out,
+            multicast_buffer_ptr,
+            buffer_ptrs_dev,
+            buffer_mnnvl,
+            buffer_flags_mnnvl,
+            nranks,
+            rank,
+            wait_for_results,
+            launch_with_pdl,
+        )
+
+    return SimpleNamespace(
+        trtllm_mnnvl_all_reduce=trtllm_mnnvl_all_reduce,
+    )
+
+
+def get_allreduce_mnnvl_workspace(
+    mapping: Mapping, dtype: torch.dtype
+) -> Tuple[McastGPUBuffer, torch.Tensor, torch.Tensor, int]:
+    force_mn = os.environ.get("TRTLLM_FORCE_MNNVL_AR", "0") == "1"
+
+    # buffer shape: [3, 2, buffer_tokens, hidden_dim]
+    stride = 3 * 2 * dtype.itemsize
+    # LCM for hidden_dim: 2048, 4096, 5120, 7168, 8192 = 286720
+    # max_num_elements must be a multiple of 286720
+    lcm_hidden_dim = 286720
+    TARGET_WORKSPACE_SIZE_BYTES = 12_000_000
+    buffer_size_in_bytes = math.ceil(
+        TARGET_WORKSPACE_SIZE_BYTES / (lcm_hidden_dim * stride)
+    ) * (lcm_hidden_dim * stride)
+    max_num_elements = buffer_size_in_bytes // stride
+
+    mcast_buffer = McastGPUBuffer(
+        buffer_size_in_bytes,
+        mapping.tp_size,
+        mapping.tp_rank,
+        torch.device("cuda", mapping.local_rank),
+        mapping.is_multi_node() or force_mn,
+    )
+
+    # Initialize the unicast buffer with -0.0
+    mcast_buffer.lamport_initialize(mapping.tp_rank)
+
+    # CPU barrier since we assume this should not be called in cuda graph
+    torch.cuda.synchronize()
+    mpi_barrier()
+
+    # This is a buffer to maintain the state of this allreduce Op
+    # Should have the same lifetime with self._buffer
+    # [Buffer_ptr, Clear_ptr, Buffer_size, atomic access counter]
+    buffer_flags = torch.tensor(
+        [0, 2, max_num_elements, 0],
+        dtype=torch.uint32,
+        device=torch.device("cuda", mapping.local_rank),
+    )
+
+    return (
+        mcast_buffer,
+        buffer_flags,
+        max_num_elements,
+    )
+
+
+def trtllm_mnnvl_all_reduce(
+    inp: torch.Tensor,
+    out: torch.Tensor,
+    multicast_buffer_ptr: int,  # Pointer address as integer
+    buffer_ptrs_dev: int,  # Pointer address as integer
+    buffer_M: int,
+    buffer_flags_mnnvl: torch.Tensor,
+    nranks: int,
+    rank: int,
+    wait_for_results: bool,
+    launch_with_pdl: bool,
+) -> None:
+    """MNNVL all-reduce operation"""
```

Comment:
I think this is the main public API right? 
This probably could deserve a complete description, including the expectation of the list of parameters and the invariants before/after.

Reply:
Thanks for pointing this out! The documentation must have been removed during my cleanup. I added an improved version now.

There is another main one: the workspace setup. It has documentation also now.